### DISCUSSION TITLE
False positive circular dependency when using Adjacent Sibling combinator (+)

### DISCUSSION THREAD

**Discussion by SharpSeeEr** [tailwindlabs/tailwindcss#7975] [Help] [answered]
**What version of Tailwind CSS are you using?**
v3.0.23

**What build tool (or framework if it abstracts the build tool) are you using?**
webpack 5.59.0
vue-cli 4.5.13
postcss 8.4.12

**What version of Node.js are you using?**
v14.17.4

**What browser are you using?**
N/A

**What operating system are you using?**
Ubuntu 21.10

**Reproduction URL**
https://play.tailwindcss.com/lK5LklDdd2?file=css

**Describe your issue**

I am getting an error during compilation:
`You cannot '@apply' the 'border-t' utility here because it creates a circular dependency.'

From the css:

```postcss
.toolbar {
  @apply flex
    items-center
    px-2
    mt-2
    py-px
    border-t
    border-b;
}

.toolbar + .section.border-t,
.toolbar + .content.border-t {
  @apply mt-0;
}
```

**Comment by adamwathan** [comment, marked as answer]
Hey! This doesn't look like a false positive, you do have a circular dependency here because of how `@apply` works — it applies _every_ instance of a class to make sure that any HTML you are refactoring behaves the way it should.

See if this comment in a previous discussion is helpful:

https://github.com/tailwindlabs/tailwindcss/issues/7373#issuecomment-1034048303

Basically the second you create a custom CSS rule that includes a Tailwind class (`.toolbar + .section.border-t` in your case), you've effectively "corrupted" that Tailwind class by hanging that extra behavior off of it.

Given this input:

```css
.toolbar + .section.border-t {
  margin: 0;
}

.foo {
  @apply border-t;
}
```

...Tailwind will generate this output:

```css
.toolbar + .section.border-t {
  margin: 0;
}

.foo {
  border-top: 1px;
}
.toolbar + .section.foo {
  margin: 0;
}
```

...because you've extended the behavior of `border-t` with your rule, and now applying `border-t` needs to try and preserve that behavior.

Consider you had this code:

```html
<style>
  .toolbar + .section.border-t {
    margin: 0;
  }
</style>

<div class="toolbar">...</div>
<div class="section border-t">...</div>
```

If you wanted to extract `border-t` to a new custom class with `@apply` (like `.foo`), it _has_ to carry through that complex rule you added, or now your design would render differently.

If `.foo` was simply `border-top: 1px`, then it wouldn't be safe to replace `border-t` in your HTML with `foo` because `border-t` has that extra behavior of removing the margin.

Hope that makes sense.

**Reply by SharpSeeEr** [reply]
@adamwathan  Thanks so much for the response! I should add that I didn't encounter this error until upgrading from tailwind v2 JIT to v3.

What I don't necessarily understand still is how I extended the behavior of `border-t`.  

```postcss
.toolbar + .section.border-t {
  @apply mt-0;
} 
```

This is saying to apply a margin-top of 0 to any element with _both_ `section` and `border-t`, not apply margin-top of 0 to any element with `section` and has a `border-top: 1`.  I'm not redefining anything about `border-t`, I'm using it exactly how it is.  

Does that make sense?  Or at the very least I hope you can see where I'm coming from, which may help you help me understand.

**Reply by SharpSeeEr** [reply]
After thinking about this more and reading [#7333 (comment)](https://github.com/tailwindlabs/tailwindcss/issues/7373#issuecomment-1034048303), this statement caught my attention:

> since we apply _all_ instances of a class

Is it done this way for simplicity?  Or some other reason?  

It still seems to me that while `.toolbar` has `@apply border-t`, `.toolbar + .section.border-t` doesn't apply to the element with the `toolbar` class, it applies to a sibling with `section border-t` applied to it.  If `.section` also had `@apply border-t`, that would definitely be a circular reference.

**Reply by adamwathan** [reply]
> What I don't necessarily understand still is how I extended the behavior of border-t

You've extended it by creating a new CSS rule that includes that selector.

Previously the only behavior of `border-t` was:

_When I add `border-t` to any element, it adds a top border._

With your custom CSS, `border-t` now does two things:

_When I add `border-t` to any element, it adds a top border._
_When I add `border-t` to an element that already has the class `content` and is next to an element with the class `toolbar`, it sets the top margin to 0._

That's what I mean when I say you've extended the behavior of `border-t`.

> Is it done this way for simplicity? Or some other reason?

It's done so that the behavior is correct when refactoring.

Like I mentioned in my earlier comment, say you have this code:

```html
<style>
  .toolbar + .section.border-t {
    margin: 0;
  }
</style>

<div class="toolbar">...</div>
<div class="section border-t bg-black">...</div>
```

The golden rule of `@apply` is that you should be able to extract a set of classes to a custom class without changing any behavior.

So now say you do this:

```html
<style>
  .toolbar + .section.border-t {
    margin: 0;
  }
  .custom-class {
    @apply border-t bg-black;
  }
</style>

<div class="toolbar">...</div>
<div class="section custom-class">...</div>
```

If we don't apply _every_ instance of `border-t`, this refactoring has now broken your design because previously the second `div` had a top margin of 0, but after the extraction it wouldn't.

This is the code you would get if we _didn't_ apply every instance:

```html
<style>
  .toolbar + .section.border-t {
    margin: 0;
  }
  .custom-class {
    border-top: 1px;
    background-color: #000;
  }
</style>

<div class="toolbar">...</div>
<div class="section custom-class">...</div>
```

So you can see why the top margin would be removed right, because there's no CSS targeting that element anymore. This makes this an unsafe refactoring, which breaks the golden rule of `@apply`.

The only way for this refactoring to be safe is for us to apply every instance of the class, giving you this result:

```html
<style>
  .toolbar + .section.border-t {
    margin: 0;
  }
  .toolbar + .section.custom-class {
    margin: 0;
  }
  .custom-class {
    border-top: 1px;
    background-color: #000;
  }
</style>

<div class="toolbar">...</div>
<div class="section custom-class">...</div>
```

That is the only way for these two chunks of HTML to behave the same in the browser:

```html
<!-- No extraction -->
<div class="toolbar">...</div>
<div class="section border-t bg-black">...</div>

<!-- With extraction -->
<div class="toolbar">...</div>
<div class="section custom-class">...</div>
```

Hope that is clear — it is absolutely 100% a Tailwind anti-pattern to write custom CSS that targets existing CSS classes. Utility classes are meant to do just a single thing, and when you write custom CSS that includes those classes in the selector, you are extending the behavior of those utilities as I've explained. This causes the ripple effect you're experiencing where it changes what happens when you `@apply` that utility, because now we have to preserve more behavior for refactoring with `@apply` to be safe.

**Reply by SharpSeeEr** [reply]
Thank you again for taking the time to offer such in depth explanations. Between your two responses I have a much better understanding, even if it isn't crystal clear at this point.   Regardless, I accept your answer because obviously you have a much more in depth knowledge and understanding of the topic.  So I accept your answer as true, and will find the rest as I go.

Also, thanks for making such a great framework!  TailwindCss is the css unicorn I wanted for the last 20 years.

(This is meant as sincere, and in no way sarcastic)
